frontend/src/components/accountPage/UserListLogic/UserListLogic.tsx

```diff
@@ -0,0 +1,206 @@
+    return `${workspaceName}-${
+      userId || Math.random().toFixed(10).replace('0.', '')
+    }`;
+  };
+
+  const handleAbort = (value: boolean) => setAbortUploading(value);
+  useEffect(() => {
+    if (!loading) {
+      setUsers(
+        data?.tenants?.items?.map(user => ({
+          key: user?.metadata?.tenantId!,
+          userid: user?.metadata?.tenantId!,
+          name: user?.spec?.firstName!,
+          surname: user?.spec?.lastName!,
+          email: user?.spec?.email!,
+          currentRole: user?.spec?.workspaces?.find(roles =>
+            workspaceName.includes(roles?.name!)
+          )?.role!,
+          workspaces:
+            user?.spec?.workspaces?.map(workspace => ({
+              role: workspace?.role! as Role,
+              name: workspace?.name! as string,
+            })) || [],
+        })) || []
+      );
+    }
+  }, [loading, data, workspaceName]);
+
+  const [applyTenantMutation] = useApplyTenantMutation();
+
+  const updateUser = async (user: UserAccountPage, newRole: Role) => {
+    try {
+      let workspaces = users
+        .find(u => u.userid === user.userid)!
+        .workspaces?.filter(w => w.name === workspaceName)
+        .map(({ name }) => ({ name, role: newRole }));
+      setLoadingSpinner(true);
+      await applyTenantMutation({
+        variables: {
+          tenantId: user.userid,
+          patchJson: getTenantPatchJson({ workspaces }),
+          manager: getManager(),
+        },
+      });
+      setUsers(
+        users.map(u =>
+          u.userid === user.userid
+            ? {
+                ...u,
+                currentRole: newRole,
+                workspaces,
+              }
+            : u
+        )
+      );
+    } catch (error) {
+      setLoadingSpinner(false);
+      return false;
+    }
+    setLoadingSpinner(false);
+    return true;
+  };
+
+  const addUser = async (usersToAdd: UserAccountPage[], workspaces: any) => {
+    try {
+      setLoadingSpinner(true);
+      setUploadedNumber(0);
+      setUploadedUserNumber(0);
+      setErrors([]);
+      let usersAdded: UserAccountPage[] = [];
+
+      for (const user of usersToAdd) {
+        if (abortUploadingRef.current) break;
+        try {
+          await applyTenantMutation({
+            variables: {
+              manager: getManager(),
+              tenantId: user.userid,
+              patchJson: getTenantPatchJson({
+                email: user.email,
+                firstName: user.name,
+                lastName: user.surname,
+                workspaces,
+              }),
+            },
+          });
+          user.workspaces?.push(workspaces);
+          setUploadedUserNumber(number => number + 1);
+          usersAdded.push(user);
+        } catch (err) {
+          setErrors(errors => [...errors, err]);
+        }
+        setUploadedNumber(number => number + 1);
+      }
+      setUsers([...users, ...usersAdded]);
+    } catch (_) {
+      // TODO: handle error when mutation fails
```

Comment:
same thing here for the error logic, maybe contact @cheina97 for further information